From f8d6ee7eaeda225f677406e31b2b1bb53ba25b9d Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Thu, 1 Oct 2026 13:48:51 +0200
Subject: [PATCH] jack: add locking around ports array and mixer list

The port array and port mix list are accessed from the data thread and
so we should protect modifications with a lock.

One problem is the port register, which might reallocate the array and
cause a race with the port->valid check and cause a crash.

Fixes #5499

diff --git a/pipewire-jack/src/pipewire-jack.c b/pipewire-jack/src/pipewire-jack.c
index 454b0d134..8cbe4b6be 100644
--- a/pipewire-jack/src/pipewire-jack.c
+++ b/pipewire-jack/src/pipewire-jack.c
@@ -710,10 +710,12 @@ static struct mix *create_mix(struct client *c, struct port *port,
 	spa_list_remove(&mix->link);
 	spa_list_append(&c->mix, &mix->link);
 
-	spa_list_append(&port->mix, &mix->port_link);
-
 	init_mix(mix, mix_id, port, peer_id);
 
+	pw_loop_lock(c->loop->loop);
+	spa_list_append(&port->mix, &mix->port_link);
+	pw_loop_unlock(c->loop->loop);
+
 	return mix;
 }
 
@@ -796,7 +798,9 @@ static struct port * alloc_port(struct client *c, enum spa_direction direction)
 
 	p->direction = direction;
 	p->emptyptr = SPA_PTR_ALIGN(p->empty, c->max_align, float);
+	pw_loop_lock(c->loop->loop);
 	p->port_id = pw_map_insert_new(&c->ports[direction], p);
+	pw_loop_unlock(c->loop->loop);
 	c->n_ports++;
 
 	pthread_mutex_lock(&c->context.lock);

